File: backend/app/api/sessoes.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from datetime import datetime, timezone
from typing import List, Optional

from app.core.database import get_db
from app.core.security import get_current_user, require_role
from app.models.user import User
from app.models.disciplina import Disciplina, DisciplinaProfessor
from app.models.sessao import Sessao, Presenca
from app.schemas.schemas import SessaoCreate, SessaoOut
from app.services.token_service import rotate_token, get_token_info

router = APIRouter()


def _sessao_out(s: Sessao, disc: Disciplina = None, checkin_count: int = 0) -> dict:
    return {
        "id": s.id,
        "disciplina_id": s.disciplina_id,
        "disciplina_nome": disc.nome if disc else None,
        "disciplina_modo": disc.modo.value if disc else None,
        "disciplina_cor": disc.cor if disc else None,
        "professor_id": s.professor_id,
        "descricao": s.descricao,
        "data": s.data,
        "token_atual": s.token_atual,
        "token_emitido_em": s.token_emitido_em,
        "ativa": s.ativa,
        "criado_em": s.criado_em,
        "encerrado_em": s.encerrado_em,
        "total_checkins": checkin_count,
    }


@router.post("/", response_model=SessaoOut)
async def create_sessao(
    body: SessaoCreate,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin", "professor")),
):
    # Validate professor has access
    if current.perfil.value == "professor":
        check = await db.execute(
            select(DisciplinaProfessor).where(
                DisciplinaProfessor.disciplina_id == body.disciplina_id,
                DisciplinaProfessor.professor_id == current.id,
            )
        )
        if not check.scalar_one_or_none():
            raise HTTPException(403, "Sem acesso a esta disciplina")

    # Deactivate any active session for this professor
    active = await db.execute(
        select(Sessao).where(Sessao.professor_id == current.id, Sessao.ativa == True)
    )
    for s in active.scalars().all():
        s.ativa = False
        s.encerrado_em = datetime.now(timezone.utc)

    data = body.data or datetime.now(timezone.utc).strftime("%Y-%m-%d")
    sessao = Sessao(
        disciplina_id=body.disciplina_id,
        professor_id=current.id,
        descricao=body.descricao,
        data=data,
    )
    db.add(sessao)
    await db.flush()

    # Generate first token
    await rotate_token(sessao, db)

    disc = await db.get(Disciplina, body.disciplina_id)
    return SessaoOut(**_sessao_out(sessao, disc))


@router.post("/{sessao_id}/rotate-token")
async def rotate(
    sessao_id: int,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin", "professor")),
):
    sessao = await db.get(Sessao, sessao_id)
    if not sessao or not sessao.ativa:
        raise HTTPException(404, "Sessão não encontrada ou encerrada")
    if sessao.professor_id != current.id and current.perfil.value != "admin":
        raise HTTPException(403, "Sem permissão")

    token = await rotate_token(sessao, db)
    return get_token_info(sessao)


@router.get("/{sessao_id}/token-info")
async def token_info(
    sessao_id: int,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin", "professor")),
):
    sessao = await db.get(Sessao, sessao_id)
    if not sessao:
        raise HTTPException(404, "Sessão não encontrada")
    return get_token_info(sessao)


@router.post("/{sessao_id}/encerrar", response_model=SessaoOut)
async def encerrar_sessao(
    sessao_id: int,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin", "professor")),
):
    sessao = await db.get(Sessao, sessao_id)
    if not sessao:
        raise HTTPException(404, "Sessão não encontrada")
    if sessao.professor_id != current.id and current.perfil.value != "admin":
        raise HTTPException(403, "Sem permissão")

    sessao.ativa = False
    sessao.encerrado_em = datetime.now(timezone.utc)
    await db.commit()
    await db.refresh(sessao)

    disc = await db.get(Disciplina, sessao.disciplina_id)
    count = await db.execute(
        select(func.count()).select_from(Presenca).where(
            Presenca.sessao_id == sessao.id, Presenca.presente == True
        )
    )
    return SessaoOut(**_sessao_out(sessao, disc, count.scalar()))


@router.get("/ativa")
async def get_active(
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin", "professor")),
):
    query = select(Sessao).where(Sessao.ativa == True)
    if current.perfil.value == "professor":
        query = query.where(Sessao.professor_id == current.id)

    result = await db.execute(query)
    sessao = result.scalar_one_or_none()
    if not sessao:
        return None

    disc = await db.get(Disciplina, sessao.disciplina_id)
    count_result = await db.execute(
        select(func.count()).select_from(Presenca).where(
            Presenca.sessao_id == sessao.id, Presenca.presente == True
        )
    )
    return SessaoOut(**_sessao_out(sessao, disc, count_result.scalar()))


@router.get("/historico", response_model=List[SessaoOut])
async def historico(
    disciplina_id: Optional[int] = None,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin", "professor")),
):
    query = select(Sessao)
    if current.perfil.value == "professor":
        query = query.where(Sessao.professor_id == current.id)
    if disciplina_id:
        query = query.where(Sessao.disciplina_id == disciplina_id)
    query = query.order_by(Sessao.data.desc(), Sessao.criado_em.desc())

    result = await db.execute(query)
    sessoes = result.scalars().all()

    out = []
    for s in sessoes:
        disc = await db.get(Disciplina, s.disciplina_id)
        count_result = await db.execute(
            select(func.count()).select_from(Presenca).where(
                Presenca.sessao_id == s.id, Presenca.presente == True
            )
        )
        out.append(SessaoOut(**_sessao_out(s, disc, count_result.scalar())))
    return out


@router.get("/{sessao_id}/presencas")
async def get_presencas(
    sessao_id: int,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(get_current_user),
):
    result = await db.execute(
        select(Presenca, User.nome).join(User, User.id == Presenca.aluno_id).where(
            Presenca.sessao_id == sessao_id
        ).order_by(User.nome)
    )
    return [
        {
            "id": p.id,
            "sessao_id": p.sessao_id,
            "aluno_id": p.aluno_id,
            "aluno_nome": nome,
            "presente": p.presente,
            "modalidade": p.modalidade.value,
            "metodo": p.metodo.value,
            "checkin_em": p.checkin_em,
        }
        for p, nome in result.all()
    ]


@router.delete("/{sessao_id}")
async def delete_sessao(
    sessao_id: int,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(require_role("admin")),
):
    sessao = await db.get(Sessao, sessao_id)
    if not sessao:
        raise HTTPException(404)
    # Delete presencas first
    await db.execute(
        Presenca.__table__.delete().where(Presenca.sessao_id == sessao_id)
    )
    await db.delete(sessao)
    await db.commit()
    return {"ok": True}

File: backend/src/routes/sessoes.js
import express from 'express';
import { prisma } from '../prisma.js';
import { requireAuth, requireRole } from '../middleware/auth.js';
import { rotateToken, getTokenInfo } from '../services/tokenService.js';

const router = express.Router();

function sessionOutput(session, disciplina, checkinCount = 0) {
  return {
    id: session.id,
    disciplina_id: session.disciplinaId,
    disciplina_nome: disciplina?.nome,
    disciplina_modo: disciplina?.modo,
    disciplina_cor: disciplina?.cor,
    professor_id: session.professorId,
    descricao: session.descricao,
    data: session.data,
    token_atual: session.tokenAtual,
    token_emitido_em: session.tokenEmitidoEm,
    ativa: session.ativa,
    criado_em: session.criadoEm,
    encerrado_em: session.encerradoEm,
    total_checkins: checkinCount,
  };
}

router.post('/', requireAuth, requireRole('admin', 'professor'), async (req, res) => {
  const current = req.user;
  const { disciplina_id, descricao, data } = req.body;
  if (!disciplina_id) {
    return res.status(400).json({ detail: 'disciplina_id é obrigatório' });
  }

  if (current.perfil === 'professor') {
    const access = await prisma.disciplinaProfessor.findFirst({
      where: {
        disciplinaId: Number(disciplina_id),
        professorId: current.id,
      },
    });
    if (!access) {
      return res.status(403).json({ detail: 'Sem acesso a esta disciplina' });
    }
  }

  await prisma.sessao.updateMany({
    where: { professorId: current.id, ativa: true },
    data: { ativa: false, encerradoEm: new Date() },
  });

  const formattedDate = data || new Date().toISOString().slice(0, 10);
  const session = await prisma.sessao.create({
    data: {
      disciplinaId: Number(disciplina_id),
      professorId: current.id,
      descricao,
      data: formattedDate,
    },
  });

  const rotated = await rotateToken(session, prisma);
  const disciplina = await prisma.disciplina.findUnique({ where: { id: Number(disciplina_id) } });
  return res.json(sessionOutput(rotated, disciplina));
});

router.post('/:sessaoId/rotate-token', requireAuth, requireRole('admin', 'professor'), async (req, res) => {
  const { sessaoId } = req.params;
  const session = await prisma.sessao.findUnique({ where: { id: Number(sessaoId) } });
  if (!session || !session.ativa) {
    return res.status(404).json({ detail: 'Sessão não encontrada ou encerrada' });
  }
  if (session.professorId !== req.user.id && req.user.perfil !== 'admin') {
    return res.status(403).json({ detail: 'Sem permissão' });
  }

  const rotated = await rotateToken(session, prisma);
  return res.json(getTokenInfo(rotated));
});

router.get('/:sessaoId/token-info', requireAuth, requireRole('admin', 'professor'), async (req, res) => {
  const { sessaoId } = req.params;
  const session = await prisma.sessao.findUnique({ where: { id: Number(sessaoId) } });
  if (!session) {
    return res.status(404).json({ detail: 'Sessão não encontrada' });
  }
  return res.json(getTokenInfo(session));
});

router.post('/:sessaoId/encerrar', requireAuth, requireRole('admin', 'professor'), async (req, res) => {
  const { sessaoId } = req.params;
  const session = await prisma.sessao.findUnique({ where: { id: Number(sessaoId) } });
  if (!session) {
    return res.status(404).json({ detail: 'Sessão não encontrada' });
  }
  if (session.professorId !== req.user.id && req.user.perfil !== 'admin') {
    return res.status(403).json({ detail: 'Sem permissão' });
  }

  const updated = await prisma.sessao.update({
    where: { id: Number(sessaoId) },
    data: { ativa: false, encerradoEm: new Date() },
  });

  const disciplina = await prisma.disciplina.findUnique({ where: { id: updated.disciplinaId } });
  const checkinCount = await prisma.presenca.count({
    where: { sessaoId: updated.id, presente: true },
  });
  return res.json(sessionOutput(updated, disciplina, checkinCount));
});

router.get('/ativa', requireAuth, requireRole('admin', 'professor'), async (req, res) => {
  const query = { ativa: true };
  if (req.user.perfil === 'professor') {
    query.professorId = req.user.id;
  }

  const session = await prisma.sessao.findFirst({ where: query });
  if (!session) {
    return res.json(null);
  }

  const disciplina = await prisma.disciplina.findUnique({ where: { id: session.disciplinaId } });
  const count = await prisma.presenca.count({
    where: { sessaoId: session.id, presente: true },
  });
  return res.json(sessionOutput(session, disciplina, count));
});

router.get('/historico', requireAuth, requireRole('admin', 'professor'), async (req, res) => {
  const { disciplina_id } = req.query;
  const where = {};
  if (req.user.perfil === 'professor') {
    where.professorId = req.user.id;
  }
  if (disciplina_id) {
    where.disciplinaId = Number(disciplina_id);
  }

  const sessions = await prisma.sessao.findMany({
    where,
    orderBy: [
      { data: 'desc' },
      { criadoEm: 'desc' },
    ],
  });

  const output = [];
  for (const session of sessions) {
    const disciplina = await prisma.disciplina.findUnique({ where: { id: session.disciplinaId } });
    const count = await prisma.presenca.count({
      where: { sessaoId: session.id, presente: true },
    });
    output.push(sessionOutput(session, disciplina, count));
  }
  return res.json(output);
});

router.get('/:sessaoId/presencas', requireAuth, async (req, res) => {
  const { sessaoId } = req.params;
  const registros = await prisma.presenca.findMany({
    where: { sessaoId: Number(sessaoId) },
    include: { aluno: true },
    orderBy: { aluno: { nome: 'asc' } },
  });

  return res.json(registros.map((p) => ({
    id: p.id,
    sessao_id: p.sessaoId,
    aluno_id: p.alunoId,
    aluno_nome: p.aluno.nome,
    presente: p.presente,
    modalidade: p.modalidade,
    metodo: p.metodo,
    checkin_em: p.checkinEm,
  })));
});

router.delete('/:sessaoId', requireAuth, requireRole('admin'), async (req, res) => {
  const { sessaoId } = req.params;
  await prisma.presenca.deleteMany({ where: { sessaoId: Number(sessaoId) } });
  await prisma.sessao.delete({ where: { id: Number(sessaoId) } });
  return res.json({ ok: true });
});

export default router;
